### ISSUE TITLE
Can GAPInfo.TestData be thread local?

### ISSUE THREAD

**Issue by alex-konovalov** [gap-system/gap#80]
An attempt to view `GAPInfo` or `GAPInfo.TestData` now fais with:
```
gap> GAPInfo.TestData;
Error, RecNames: <rec> must be a record (not a thread-local record) in
  rec( 
  names := List( RecNames( record ) )
     ; on line 296 of file /Users/alexk/GITREPS/hpcgap/lib/record.g
```
Indeed, `GAPInfo` is an atomic record and `TestData` is a thread local record. Perhaps `TestData` should not be kept in `GAPInfo` any more.

**Comment by fingolfin** [comment]
Note that semigraphs and digraphs access `GAPInfo.TestData`, so if we change this, we should adjust them, too.

Perhaps we can also just get rid of it completely; we'd have to learn why these packages access it; then perhaps we can find a way to provide the same functionality to them in a cleaner way...?

@james-d-mitchell perhaps you can shine some light on this?

**Comment by alex-konovalov** [comment]
It's not documented, but `system.g` contains unlinked mansection which says that this is a mutable record, which keeps some data used by `START_TEST` and `STOP_TEST`. The data refer to a particular run of the test, so they have to be thread local to allow to run the same test in concurrent threads.

**Comment by james-d-mitchell** [comment]
@fingolfin in the Semigroups (!) package we are using `GAPInfo.TestData` to time how long the individual manual examples take to run. In Digraphs we were using this to display the amount of time a test file took to run in milliseconds, when GAP still displayed this in GAPstones. We can easily remove the use of `GAPInfo.TestData` from Digraphs, and also from the Semigroups package if this is desirable.

**Comment by fingolfin** [comment]
@james-d-mitchell oops, stupid braino on the package name, now corrected :-).

If getting this data is useful for you, then I'd rather think about coming up a way to give it to you "officially", i.e., via a documented mechanism. I do not want to just document `GAPInfo.TestData`, though, as I think it's a bad idea to "return" results of a function by storing it into some global variable (the problems this causes in a multi threaded environment are just one reason for that).

Instead, e.g. `Test` could return a record with all that data. Of course `Test` right now returns a boolean, but I am not sure anybody uses that -- but if they are, we could introduce an option for `Test` which toggles the return value; e.g. `return_extended_data:=true`.

**Comment by james-d-mitchell** [comment]
Agreed, I think an official route for obtaining this information would be useful, and what we have in Semigroups and Digraphs was just temporary hacks. The `return_extended_data` sounds like a good idea. How would you envision doing this?

**Comment by james-d-mitchell** [comment]
I'm just updating Digraphs to remove the use of `GAPInfo.TestData`.

**Comment by alex-konovalov** [comment]
For the record, `GAPInfo.TestData` is not in use by packages any more.

**Comment by fingolfin** [comment]
As far as I can tell, semigroups (version 3) still uses it, albeit only in one spot now:
```
gap/tools/utils.gi:172:  record.elapsed_last_test := Runtime() - GAPInfo.TestData.START_TIME;
```

Anyway, I am sure @james-d-mitchell would be quick to adapt if we were to finally move `GAPInfo.TestData` out of there and into a thread local record, so I am not really concerned about this usage :-).

And there is of course also that idea of returning additional data in `Test`. I have no plan to work on that, but pull requests would be welcome. Basically, just add a new entry for the options record (perhaps `returnTimings` or `returnExtendedResult` would be a better name, fitting into our usual camelCase naming convention). It should default to `false`, and be documented as such. Then, if it is set to `true`, return a record, with at least the component `result` (or `success`, or whatever) bound to the boolean value normally returned by `Test`; any other components can then be added. E.g. `runtime` or `usedMemory` or whatever else.

Naturally, `TestDirectory` could receive a similar treatment.

**Comment by james-d-mitchell** [comment]
I'm happy to replace this, if I know what to replace it with :)

**Comment by alex-konovalov** [comment]
Semigroups is not using it any more: https://github.com/gap-packages/Semigroups/commit/63a91963cc1bd8099849a1bac68540e8ef3130ff#diff-f0614c52a12ccef4a6533853944d21d9

**Comment by fingolfin** [comment]
Just had a fresh look at this. It seems to me that the real issue here is that printing `GAPInfo.TestData` does not work in HPC-GAP because we end up trying to invoke `RecNames` on it, which is not implemented. But actually there is code for printing these, it just is restricted to `PrintObj`:
```
gap> GAPInfo.TestData.foo := 42;;
gap> Print(GAPInfo.TestData, "\n");
rec( foo := 42 )
gap> Display(GAPInfo.TestData);
rec( foo := 42 )
gap> ViewObj(GAPInfo.TestData);
Error, REC_NAMES: <rec> must be a record (not a thread-local record) in
  RecNames( record
```
So there are (at least) three ways the original issue could be resolved (and we may wish to do all three, as this would resolve other issues; or perhaps at least 2+3):
1. we could add a `ViewObj` method for thread local records (which might just do the same as the `PrintObj` method)
2. we could implement `RecNames` for thread local records (that would likely be beneficial anyway?)
3. we could replace the "thread-local record" `GAPInfo.TestData` by a "record that is thread-local", so e.g. `BindThreadLocal("GAPInfo_TestData", rec());`
